---
title: "EML R package overview"
author: "Matt"
date: "`r format(Sys.time(), '%d %M, %Y')`"
output: html_document
bibliography: bibliography.bib
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(knitr.duplicate.label = "allow")

```

****
*This is a "Living Document" which means that it will be updated periodically and that anyone can contribute to the document. If you wish to contribute please use a GitHub pull request at the https://github.com/LivingNorway/LivNoR repository*

****

```{r libraries, include=FALSE, message=FALSE, warning=FALSE}
#CRAN (25 documents 7 packages)
library(tidyverse)
library(EML)
library(emld)
library(emldown)
library(RNeXML)
library(datapack)
library(rbefdata)
library(geometa)

#GitHub (15 repositories, 6 R language)

#devtools::install_github("isteves/dataspice")#fails
#devtools::install_github("cboettig/eml2")#fails
#devtools::install_github("earnaud/MetaShARK-v2", dependencies=TRUE)
#devtools::install_github("atn38/MetaInbase", dependencies=TRUE)
#devtools::install_github("EDIorg/EMLassemblyline")
#devtools::install_github("BLE-LTER/MetaEgress")

library(MetaEgress)
library(EMLassemblyline)
library(MetaInbase)
library(MetaShARK)


library(jsonlite)
library(magrittr) # for pipes
library(jqr)      # for JQ examples only
library(rdflib)   # for RDf examples only

```

# R packages that interact with EML 

## Introduction

EML (Ecological Metadata Langauage; [@EML_2019]) is a version of the machine and human readable XML (Extensible Markup Language) which is designed to facilitate data documentation (i.e. metadata) to ease open data and sharing. EML consists of "modules" that allow the description of data attributes (e.g. spatial, temporal, taxonomic extent).

GBIF (the Global Biodiversity Information Facility), which is an international network and research infrastructure, provides open access to data about biodiversity. The GBIF metadata profile is based on EML and as such anyone interacting with GBIF needs to be able to read, write or visualise data in the EML format. 

Writing directly in EML can be a technical barrier to many researchers who, in ecology, may only have experience in the R langauge which has a very different syntax and data structure. Several R packages have functions to read, write or interact with EML files and here we aim to identify those that will be useful for researchers working with ecological data.  

### Aim
To identify R packages and functions that can read, write or interact/visualise data in the EML format 

#### Objectives
1) To review the available packages on CRAN and on GitHub that have functions to interact with EML
2) To identify the main functions from these packages that allow reading, writing and visualising data in EML format

## Methods

We searched for R packages on CRAN (The Comprehensive R Archive Network - https://cran.r-project.org/) using the search function which is provided by Google. We used the search term "Ecological Metadata Language site:r-project.org". We repeated the search for "Ecological Metadata Language" on GitHub and identified repositories which used the R langauge. We identified all packages returned through our search that were on CRAN or on GitHub and that consisted of functions that are aimed at interacting (either reading, writing or viewing) with the Ecological Metadata Langauge (EML).

## Results

The search on Google (via CRAN) returned 25 hits. Seven of these were R packages. The search on GitHub returned 15 repositories of which 6 were in the R language. Two of these returned errors when we attempted to download them in RStudio using the devtools::install_github() function. 
We identifed 11 R packages (7 from CRAN and 4 from GitHub) that were active and explicitly stated that functions were used for interacting (either reading, writing or viewing) Ecological Metadata Langauge (EML). 

```{r table of packages, echo=FALSE, message=FALSE, warning=FALSE}
library(knitr)
tb<-data.frame("Package"=c("EML","emld", "emldown", "RNeXML", "datapack", "rbefdata", "geometa", "MetaEgress", "EMLassemblyline", "MetaInbase", "MetaShARK"), "Package_description"=c(packageDescription(pkg = "EML")$Description, packageDescription(pkg = "emld")$Description, packageDescription(pkg = "emldown")$Description, packageDescription(pkg = "RNeXML")$Description,packageDescription(pkg = "datapack")$Description,packageDescription(pkg = "rbefdata")$Description,packageDescription(pkg = "geometa")$Description, packageDescription(pkg = "MetaEgress")$Description, packageDescription(pkg = "EMLassemblyline")$Description, packageDescription(pkg = "MetaInbase")$Description,packageDescription(pkg = "MetaShARK")$Description))

tb %>% 
  mutate(Package_description=gsub("\n", "", Package_description )) %>% knitr::kable()
```


*****

```{r, echo=FALSE}
# library(collapsibleTree)
# eml<-data.frame("Level1"=c("dataset",
#                                     "creator",
#                                     "title",
#                                     "publisher",
#                                     "pubDate",
#                                     "keywords",
#                                     "abstract",
#                                     "intellectualRights",
#                                     "contact",
#                                     "methods",
#                                     "coverage",
#                                     "coverage",
#                                     "coverage",
#                                     "dataTable",
#                                     "dataTable",
#                                     "dataTable",
#                                     "dataTable"), "Level2"=c(NA,
#                                                            NA,
#                                                            NA,
#                                                            NA,
#                                                            NA,
#                                                            NA,
#                                                            NA,
#                                                            NA,
#                                                            NA,
#                                                            NA,
#                                                            "geographicCoverage",
#                                                            "temporalCoverage",
#                                                            "taxonomicCoverage",
#                                                            "entityName",
#                                                            "entityDescription",
#                                                            "physical",
#                                                            "attributeList"))
#collapsibleTree(eml, hierarchy =c("Level1", "Level2"))
```



#### emld package

The *emld* package is closely related to the EML package and its functions are heavily relied upon in the latest version of *EML*. 

*emld* has an advantage over *EML* where there are large, highly nested EML files. It can flattern EML in to common R formats that can be manipulated in R. 

##### Read EML

```{r}
# Example from https://github.com/ropensci/emld
f <- system.file("extdata/example.xml", package="emld")
eml <- emld::as_emld(f)
eml$dataset$title
```

##### Write EML

*emld* objects are nested lists so to write EML in *emld* you can create a list.

```{r}
# Example from https://github.com/ropensci/emld

me <- list(individualName = list(givenName = "Joe", surName = "Bloggs"))

eml <- list(dataset = list(
              title = "The biggest fish in the sea",
              contact = me,
              creator = me),
              system = "doi",
              packageId = "10.xxx")

ex.xml <- tempfile("ex", fileext = ".xml") # use your preferred file path

as_xml(eml, ex.xml)

```
*emld::as_xml()* reorders the list if neccessary to ensure that it matches the EML required format. You can use *emld::eml_validate()* to check that the EML has been written correctly and is a valid EML file.

```{r}
eml_validate(ex.xml)
```


#### EML package

The *EML* package uses *emld* as a basis for many of its functions. 

##### Read EML

The read_eml() function reads eml files in to the R session. 


```{r}
f <- system.file("tests", emld::eml_version(), "eml-datasetWithUnits.xml", package = "emld")
eml <- read_eml(f)
eml$dataset$title
```

##### Write EML

Several functions (with the prefix *set_*) allow the user to import attributes of the dataset file (e.g. the methods) as word or Markdown documents and merge them in to a single EML file (see https://docs.ropensci.org/EML/ for a detailed tutorial).

##### Interact with EML

*EML* has several functions (with the prefix *get_*) which allow a user to extract elements of an XML file. 


```{r}
f <- system.file("tests", emld::eml_version(), 
  "eml-datasetWithAttributelevelMethods.xml", package = "emld")
eml <- read_eml(f)
get_attributes(eml$dataset$dataTable$attributeList)

```

## emldown package

*emldown* is a package designed to allow visualisation and interaction with EML. 

##### Interact with EML

```{r emldown}
library("emldown")
render_eml(system.file("extdata", "hampton_underice.xml", package = "emldown")) 

#This creates a human readable version of the EML file in your browser

```


### RNeXML package

*RNeXML* is a package for reading and writing phylogenetic, character and trait metadata focussed on taxonomy. The package stands parallel to *EML* and *emd* using similar functions but is focused on a different (but related) standard NeXML rather than EML per se.


### datapack package

The *datapack* package provides functions for collating multiple data and metadata objects of different types into a bundle that can be transported and loaded using a single composite file. It is primarily meant as a container to bundle together files for transport to or from DataONE data repositories. Metadata in the form of EML can be attached to the databundle. 

```{r datapack, echo=FALSE}
# metadataFile <- system.file("extdata/sample-eml.xml", package="datapack")
# metadataId <- "metadataId"
# metadataObj <- new("DataObject", id=metadataId, format="eml://ecoinformatics.org/eml-2.1.0", file=metadataFile)
# metadataObj


```

### rbefdata package

Similaty to the *datapack* package *rbefdata* links to the BEF data portal (https://fundiv.befdata.biow.uni-leipzig.de/). The function *rbefdata::bef.portal.get.metadata()* extracts the metadata from a file that the user has downloaded from the BEF data portal.

### geometa package

*geometa* provides functions for reading and writing geographic metadata. It suggests *EML* and *emld*. It can convert metadat to and from EML using the *geometa::convert_metadata()* function. 

### MetaEgress package

*MetaEgress* is designed to create EML for Long Term Ecological Research metabase ( https://github.com/lter/LTER-core-metabase). Validation of EML is done through *emld::eml_validate()*

### EMLassemblyline package

*EMLassemblyline* is a metadata builder that has functions to autoextract metadata. The package is centered around a "data package" that is a collection of data objects and metadata. The templating functions allow the development of metadata templates that the user can add information to. *EMLassemblyline* also has a "living data" set of functions that allow ongoing data collection to be published at regular periods. See https://ediorg.github.io/EMLassemblyline/index.html for examples of the workflow.

```{r EMLassembly, echo=FALSE}
#report9$SummaryReporter$get_summary_view()


```

### MetaInbase package

*MetaInbase* has functions to convert EML to tables. It appears to be in an early stage of development


## MetaShARK package

*MetaShARK* (Metadata Shiny Automated Resource & Knowledge) is a R shiny app allowing the user to get information about EML and to fill in metadata for datasets according to this standard. This package has developed a user-friendly Shiny App for researchers not familiar with EML standards.


```{r MetaShARK, echo=FALSE}
#report11$SummaryReporter$get_summary_view()


```





```{r, echo=FALSE}
# Dependancies
#library(tidyverse)
# 
# length(report1$DependencyReporter$nodes$node)#48
# length(report2$DependencyReporter$nodes$node)#13
# length(report3$DependencyReporter$nodes$node)#79
# length(report4$DependencyReporter$nodes$node)#49
# length(report5$DependencyReporter$nodes$node)#39
# length(report6$DependencyReporter$nodes$node)#17
# length(report7$DependencyReporter$nodes$node)#16
# length(report8$DependencyReporter$nodes$node)#58
# length(report9$DependencyReporter$nodes$node)#103
# #length(report10$DependencyReporter$nodes$node)#No Dependency
# length(report11$DependencyReporter$nodes$node)#170
# 
# PkgName=c(rep("EML",48), rep("emld",13), rep("emldown",79), rep("RNeXML", 49),
#           rep("datapack", 39), rep("rbefdata", 17), rep("geometa", 16), rep("MetaEgress", 58), 
#           rep("EMLassemblyline", 103), rep("MetaShARK", 170))
# 
# Dependency=c(report1$DependencyReporter$nodes$node,
#                   report2$DependencyReporter$nodes$node,
#                   report3$DependencyReporter$nodes$node,
#                   report4$DependencyReporter$nodes$node,
#                   report5$DependencyReporter$nodes$node,
#                   report6$DependencyReporter$nodes$node,
#                   report7$DependencyReporter$nodes$node,
#                   report8$DependencyReporter$nodes$node,
#              report9$DependencyReporter$nodes$node,
#              #report10$DependencyReporter$nodes$node,
#              report11$DependencyReporter$nodes$node)
# 
# dtab=data.frame(PkgName,Dependency)
# dtab$PkgName<-as.character(dtab$PkgName)
# dtab$Dependency<-as.character(dtab$Dependency)
# 
# 
# dtab=dtab %>% 
#   filter(!PkgName==Dependency)
# dtab %>% 
#   select(Dependency) %>% 
#   group_by(Dependency) %>% 
#   tally() %>% 
#   arrange(desc(n)) %>% 
#   top_n(.,15) %>% 
#   ggplot(aes(reorder(Dependency,n),n, fill="red"))+
#   geom_bar(stat="Identity")+
#   coord_flip()
# 

```


```{r open reports, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
# 
# library(pkgnet)
# report1 <- CreatePackageReport(pkg_name = "EML", report_path ="C:/Users/matthew.grainger/Documents/Projects_in_development/LivNoR/Pkgnet_out/EML.html")
# 
# report2 <- CreatePackageReport(pkg_name = "emld", report_path ="C:/Users/matthew.grainger/Documents/Projects_in_development/LivNoR/Pkgnet_out/emld.html")
# 
# report3 <- CreatePackageReport(pkg_name = "emldown",
#                                report_path ="C:/Users/matthew.grainger/Documents/Projects_in_development/LivNoR/Pkgnet_out/emldown.html")
# 
# report4 <- CreatePackageReport(pkg_name = "RNeXML",
#                                report_path ="C:/Users/matthew.grainger/Documents/Projects_in_development/LivNoR/Pkgnet_out/RNeXML.html")
# report5 <- CreatePackageReport(pkg_name = "datapack", report_path ="C:/Users/matthew.grainger/Documents/Projects_in_development/LivNoR/Pkgnet_out/datapack.html")
# report6 <- CreatePackageReport(pkg_name = "rbefdata",
#                                report_path ="C:/Users/matthew.grainger/Documents/Projects_in_development/LivNoR/Pkgnet_out/rbefdata.html")
# report7 <- CreatePackageReport(pkg_name = "geometa",
#                                report_path ="C:/Users/matthew.grainger/Documents/Projects_in_development/LivNoR/Pkgnet_out/geometa.html")
# report8 <- CreatePackageReport(pkg_name = "MetaEgress",report_path ="C:/Users/matthew.grainger/Documents/Projects_in_development/LivNoR/Pkgnet_out/MetaEgress.html")
# report9 <- CreatePackageReport(pkg_name = "EMLassemblyline"
#                                , report_path ="C:/Users/matthew.grainger/Documents/Projects_in_development/LivNoR/Pkgnet_out/EMLassemblyline.html")
# report10 <- CreatePackageReport(pkg_name = "MetaInbase", report_path ="C:/Users/matthew.grainger/Documents/Projects_in_development/LivNoR/Pkgnet_out/MetaInbase.html")#No dependancies
# report11 <- CreatePackageReport(pkg_name = "MetaShARK", report_path ="C:/Users/matthew.grainger/Documents/Projects_in_development/LivNoR/Pkgnet_out/MetaShARK.html")
# 

```




# References {-}



<div id="refs"></div>



\newpage 